src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -158,7 +161,15 @@ public virtual void OnSocketClosed()
         // Called on Libuv thread
         public void Tick()
         {
-            _frame.Tick();
+            if (_timeoutEnabled)
+            {
+                if (_secondsSinceLastConnectionEvent > ServerOptions.Limits.ConnectionTimeout.TotalSeconds)
+                {
+                    StopAsync();
+                }
+
+                _secondsSinceLastConnectionEvent++;
```

Comment:
In this case I would make a few changes:

* Track the time between connection events as milliseconds rather than seconds.
* Make the millseconds between ticks public rather than private (`KestrelThread._heartbeatMilliseconds`).
* In `Tick()`, increment the elapsed milliseconds by the public constant.